libecho/src/AudioStream.h

```diff
@@ -0,0 +1,127 @@
+#ifndef ECHOCONNECT_AUDIOSTREAM_H
+#define ECHOCONNECT_AUDIOSTREAM_H
+
+#include <QtCore/QThread>
+#include <QtMultimedia/QAudioFormat>
+#include <QtMultimedia/QAudio>
+#include <QtMultimedia/QAudioInput>
+#include <QDebug>
+#include <QMetaMethod>
+
+constexpr int NOTIFY_INTERVAL = 32;
+constexpr double VOLUME = 1.0;
+
+struct AudioStreamInfo {
+    int bufferSize;
+    int notifyInterval;
+    int periodSize;
+    qreal volume;
+};
+
+
+class AudioStreamSignalsAndSlots : public QThread {
+protected slots:
+    virtual void handleStateChanged_slot(QAudio::State newState) = 0;
+
+    virtual void handleNotify_slot() = 0;
+};
+
+template<typename StreamType>
+class AudioStream : public AudioStreamSignalsAndSlots {
+public:
+    using StatusType = std::pair<int, qint64>;
+    explicit AudioStream(const QAudioFormat& format) : format(format) {
+        qDebug() << "Constructing AudioStream. Thread:" << thread();
+        this->moveToThread(this);
+        QThread::start();
+        waitForTick();
+    }
+
+    ~AudioStream() override {
+        qDebug() << "Destroying AudioInput";
+        this->quit();
+        this->wait();
+        delete qStream;
+    }
+
+    void run() override {
```

Comment:
If you run it multiple times you have a memleak - so make sure it is called only once as you said so!

Reply:
That is not an easy task, and error is not critical / workstopper. I'll do that in other PR